src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs

```diff
@@ -0,0 +1,392 @@
+                #endif
+                if (args.Length != parameters.Length)
+                  args = new object[] { item };
+              }
+              else if (item is Array)
+              {
+                Array array = item as Array;
+
+                #if NETCF
+                if (array.Rank == 1 && (method.IsGenericMethodDefinition || array.Length == parameters.Length))
+                  #else
+                  if (array.Rank == 1 && array.Length == parameters.Length)
+                    #endif
+                {
+                  args = new object[array.Length];
+                  for (int i = 0; i < array.Length; i++)
+                    args[i] = array.GetValue(i);
+                  #if NETCF
+                  if (method.IsGenericMethodDefinition)
+                  {
+                    var mi = method.MakeGenericMethodEx(args);
+
+                    if (mi == null || array.Length != mi.GetParameters().Length)
+                      args = new object[] {item};
+                  }
+                  #endif
+                }
+                else
+                {
+                  args = new object[] { item };
+                }
+              }
+              else
+              {
+                args = new object[] { item };
+              }
+
+              parms = new TestCaseParameters(args);
+            }
+
+            data.Add(parms);
+          }
+        }
+      }
+      catch (Exception ex)
+      {
+        data.Clear();
+        data.Add(new TestCaseParameters(ex));
+      }
+
+      return data;
+    }
+    
+    #endregion
+    
+    #region ITestBuilder Members
+
+    /// <summary>
+    /// Construct one or more TestMethods from a given MethodInfo,
+    /// using available parameter data.
+    /// </summary>
+    /// <param name="method">The MethodInfo for which tests are to be constructed.</param>
+    /// <param name="suite">The suite to which the tests will be added.</param>
+    /// <returns>One or more TestMethods</returns>
+    public IEnumerable<TestMethod> BuildFrom(IMethodInfo method, Test suite)
+    {
+      foreach (TestCaseParameters tcp in GetTestCasesFor(method))
+        yield return _builder.BuildTestMethod(method, suite, tcp);
+    }
+    
+    #endregion
+
+    #region Helper Methods
+
+    private IEnumerable GetDataFromSource(IMethodInfo method){
+      
+      #region Error Checking
+      
+      // file name sources
+      if (DataType == DataSourceType.Csv || DataType == DataSourceType.Xml){
+        if (!File.Exists(SourceName)){
+          string errMsg = String.Format("Unable to find source: {0}", SourceName);
+          throw new FileNotFoundException(errMsg);
+        }
+      }
+      
+      // TODO: Database file sources
+      
+      #endregion
+      
+      /*
+       * Get data depending on the source. I catch the errors here and then marshall
+       * any known errors.
+       */
+      
+      try {
+        
+        if (DataType == DataSourceType.Csv){
```

Comment:
You should really have some sort of data source factory class that constructs a data source based on the enum type and returns a base DataSource that you call GetData on. These higher level classes should not be so tightly bound to the underlying data sources. Adding a new data source should be as simple as deriving a new DataSource class, implementing its GetData method and then adding it to the enum and factory. The factory can also be tested to ensure that it handles every enum value. That is difficult to do in this class because it is so tightly coupled with so many things.